### PR TITLE
feat(auth): add scopes to access keys creation

### PR DESCRIPTION
**Pull request by anastasia-nesterenko**
## Summary
Access keys created via `nemo auth access-keys create` can now be restricted to specific services (`--scope`), granted workspace membership at creation time (`--workspace`, replacing a separate `nemo workspaces members create` call), and rotated by JTI (`--rotate`) for personal keys, since `--service-account` is optional.

## Changes
- Add `scope`, `rotates`, and `workspaces` to access-key creation, metadata, JWT claims, storage, API responses, and OpenAPI schemas.
- Add `--scope`, `--rotate`, and repeatable `--workspace` CLI options. `--service-account` remains optional, and invalid values return clear CLI errors.
- Support personal-key rotation by JTI and preserve the original key’s scope during rotation.
- Prevent scoped keys, including PlatformAdmin keys, from creating broader keys.
- Require `entities` or `platform` scope when a restricted key grants workspace membership.
- Reject duplicate workspace grants and safely restore previous memberships when creation or rotation fails.
- Add regression tests for validation, scope enforcement, rotation, workspace grants, and rollback behavior.
- Update the generated API, SDK, CLI reference, and authentication documentation.

```
nemo auth access-keys create \
  --name "intake-worker" \
  --service-account intake-bot \
  --scope intake,entities \
  --workspace ml-team:Editor \
  --expires-in 86400

# Create a personal (user-bound) key, unscoped, multiple workspace grants
nemo auth access-keys create --name "my-laptop" --workspace ml-team --workspace analytics:Viewer,Editor

# Rotate a personal key: mints a successor and revokes the old one (jti must be one you own, no service account)
nemo auth access-keys rotate ak_1234567890abcdef1234567890abcdef

# Other lifecycle commands
nemo auth access-keys list
nemo auth access-keys suspend <jti>
nemo auth access-keys unsuspend <jti>
nemo auth access-keys revoke <jti>
```

## Type of Change

- [x] Code change (feature, bug fix, or refactor)
- [x] Code change with documentation updates
- [ ] Documentation only
- [ ] Contributor tooling or automation
- [ ] CI, build, or test infrastructure

## Quality Gates

- [x] Tests added or updated for changed behavior
- [ ] Existing tests cover changed behavior — justification:
- [ ] Tests not applicable — justification:
- [x] Documentation updated for user-visible behavior
- [ ] Documentation not applicable — justification:

## Verification

- [x] Pull request title follows the repository's Conventional Commit format — suggest `feat(auth): add service-scoped access keys`.
- [x] Every commit includes an appropriate `Signed-off-by:` trailer
- [ ] `uv run pre-commit run -a` passes, or any blocked checks are identified below — full repo-wide run not executed this session; targeted lint/type/test checks below all passed.
- [x] Targeted tests pass, or tests are marked not applicable above
- [x] No secrets, API keys, or credentials are included

## New Features
- Create access keys with service scopes, rotation targets, and workspace memberships.
- Access-key listings and responses display assigned scopes.
- Scoped keys preserve or inherit scope during rotation, with workspace memberships provisioned automatically.

## Bug Fixes
- Restricted keys cannot perform actions outside their permitted services.
- Platform administrators with restricted scopes are subject to scope enforcement.
- Invalid scopes, workspaces, roles, and duplicate workspace grants are rejected.

## Documentation
- Updated authentication guidance and CLI reference for scoped access-key options and administration requirements.

### CODE REVIEW

File: services/core/auth/src/nmp/core/auth/app/access_keys.py [hunk 1]
```diff
@@ -660,8 +703,83 @@ async def create_async(
...
+        # Provision the new key's requested access before revoking its predecessor, so a key
+        # is never left with less access than it had before rotation started.
+        try:
+            await self._grant_workspace_memberships(key, request.workspaces)
+        except Exception:
+            # A key whose requested workspace access could not be provisioned is worse than
+            # useless: the caller would believe --workspace replaced their separate
+            # `nemo workspaces members create` call when it silently didn't. Revoke it and
+            # surface the failure instead of returning a 200 with incomplete access.
+            await self._revoke_after_failed_creation(key.jti, "workspace_grant_failed")
```
Comment (ironcommit) [inline services/core/auth/src/nmp/core/auth/app/access_keys.py:740]: **Finding 1: Major Partial Workspace Grants Persist After Failed Create Or Rotation**

**Location:** `services/core/auth/src/nmp/core/auth/app/access_keys.py:708`
**Affected lines:** `708-737`, `755-781`
**Severity:** Major

---

**Evidence:** The new create flow grants workspace memberships after the access-key lifecycle record is persisted. If any later workspace grant fails, the compensation only revokes the new access key. If `rotates` is also set and predecessor revocation fails, the compensation again only revokes the new key. `_grant_workspace_memberships` does not track grants that already succeeded and does not restore the previous membership state.

**Impact:** Workspace membership is principal-wide, not key-local. A request such as `--workspace team-a --workspace missing` can fail and revoke the new key while leaving `team-a` membership active. For a personal key rotation, if workspace grants succeed and predecessor revocation fails, the old key stays active and now authenticates as the same principal that just received the new workspace roles. That violates the fail-closed intent in the comments and can leave extra access behind after a failed API call.

**Reviewed diff:**

```diff
@@
+        try:
+            await self._grant_workspace_memberships(key, request.workspaces)
+        except Exception:
+            # A key whose requested workspace access could not be provisioned is worse than
+            # useless: the caller would believe --workspace replaced their separate
+            # `nemo workspaces members create` call when it silently didn't. Revoke it and
+            # surface the failure instead of returning a 200 with incomplete access.
+            await self._revoke_after_failed_creation(key.jti, "workspace_grant_failed")
+            raise
+        if request.rotates is not None:
+            try:
+                await self.revoke_async(request.rotates)
+            except Exception:
+                ...
+                await self._revoke_after_failed_creation(key.jti, "rotation_revoke_failed")
+                raise
@@
+        for grant in workspaces or []:
+            try:
+                await self._workspaces_client.create_workspace_member(
+                    workspace=grant.workspace,
+                    body=CreateWorkspaceMemberRequest(
+                        principal=key.principal,
+                        roles=grant.roles,
+                    ),
+                )
```

**Suggested fix:** Make workspace provisioning transactional enough for this workflow. Before applying grants, snapshot each target principal's current roles per workspace, or preflight all target workspaces and authorization before mutating. If a later grant or predecessor revoke fails, restore the original memberships precisely: update back to the prior role set, delete only memberships created from nothing, and do not remove roles that existed before this request. Add tests for a multi-workspace partial failure and for `workspaces + rotates` where predecessor revocation fails, asserting both the new key status and the restored workspace membership state.

---

<details>
<summary><strong>Agent fix prompt</strong></summary>

```text
Target: `services/core/auth/src/nmp/core/auth/app/access_keys.py:708`
Affected lines: `708-737`, `755-781`

You are working in this repository. Fix the rollback gap in `services/core/auth/src/nmp/core/auth/app/access_keys.py` around workspace grants during access-key creation.

Problem:
Workspace grants are principal-wide. When a create request fails after one or more grants have succeeded, or when `rotates` fails after workspace grants succeeded, the code only revokes the new access key and leaves workspace memberships behind.

Required changes:
- Snapshot or preflight requested workspace memberships before mutation.
- On any later failure, restore the original roles for each workspace/principal that this request changed, without deleting pre-existing roles.
- Add focused tests covering partial multi-workspace grant failure and `workspaces + rotates` predecessor-revoke failure.

Constraints:
- Preserve existing project conventions.
- Keep the fix scoped to access-key create/rotation workspace provisioning.
```

</details>

File: openapi/openapi.yaml [hunk 2]
```diff
@@ -8342,6 +8342,30 @@ components:
           maxLength: 240
           minLength: 1
           pattern: ^[a-zA-Z0-9][a-zA-Z0-9._+/-]*$
+        scope:
```
Comment (ironcommit) [inline openapi/openapi.yaml:8429]: **Finding 2: Minor OpenAPI Allows Values The Server Rejects**

**Location:** `openapi/openapi.yaml:8343`
**Affected lines:** `8343-8367`, `8702-8715`
**Severity:** Minor

---

**Evidence:** `nullable: true` only allows the whole field to be omitted or sent as `null`; it does not describe the valid contents when the field is non-null. The server-side contract in `packages/nemo_platform_plugin/src/nemo_platform_plugin/auth/access_keys/types.py` is:

- `scope: list[str] | None`: omitted or `null` means "unscoped key" and is valid. If `scope` is present and non-null, it must be a non-empty array. Each service name is trimmed, must remain non-empty, and must not contain whitespace or `:` because service names are serialized into token scopes such as `intake:read intake:write`.
- `workspaces: list[AccessKeyWorkspaceGrant] | None`: omitted or `null` means "no workspace grants" and is valid. If `workspaces` is present and non-null, each item is a grant object, not a plain string.
- `AccessKeyWorkspaceGrant.workspace: str`: the workspace name is trimmed and must remain non-empty.
- `AccessKeyWorkspaceGrant.roles: list[str]`: omitting `roles` defaults to `["Editor"]`. If `roles` is provided, the validator strips/drops blank entries and rejects the grant if no non-blank role remains.

Concrete examples:

```jsonc
{}                                      // valid: scope/workspaces omitted
{"scope": null}                         // valid: unscoped key
{"scope": ["intake", "entities"]}       // valid
{"scope": []}                           // invalid: non-null scope must not be empty
{"scope": [""]}                         // invalid: service name blank after trim
{"scope": ["intake read"]}              // invalid: whitespace corrupts scope-claim serialization
{"scope": ["intake:read"]}              // invalid: ':' corrupts scope-claim serialization
{"workspaces": null}                    // valid: no workspace grants
{"workspaces": [{"workspace": "ml"}]}   // valid: roles default to ["Editor"]
{"workspaces": [{"workspace": ""}]}     // invalid: blank workspace
{"workspaces": [{"workspace": "ml", "roles": []}]} // invalid: no non-blank role
```

The generated OpenAPI schemas added in this PR only say `items: type: string` for `scope` and `roles`, and plain `type: string` for `workspace`; they do not expose the non-null array constraints, service-name item constraints, workspace non-blank constraint, role "at least one non-blank value" rule, or the `Editor` default.

**Impact:** The published REST contract and generated clients document requests that the server rejects. SDK consumers and users of the docs/API reference cannot see the real validation rules or the default role, and future generated clients may fail to validate locally or present misleading forms.

**Reviewed diff:**

```diff
@@
+        scope:
+          title: Scope
+          description: Optional service names that restrict this key to read and write
+            access for those services.
+          nullable: true
+          items:
+            type: string
+          type: array
@@
+    AccessKeyWorkspaceGrant:
+      properties:
+        workspace:
+          type: string
+          title: Workspace
+        roles:
+          items:
+            type: string
+          type: array
+          title: Roles
```

**Suggested fix:** Put the validation in the schema source, then regenerate all three OpenAPI files. Keep `scope` nullable to represent omitted/null/unscoped, but when it is non-null, express it as a non-empty array (`minItems: 1`) whose items match the service-name rules, for example non-blank with no whitespace or `:`. Keep `workspaces` nullable to represent no workspace grants, but require each grant object's `workspace` field to be a non-blank string. Express `roles` as optional/defaulting to `["Editor"]`; when provided, it should have at least one non-blank role. If the API intentionally accepts and normalizes blank role entries, model that with `contains`/`minContains`; if client-side rejection is preferred, use an item pattern. Extend `test_access_key_lifecycle_openapi_documents_error_responses` or a focused schema test to assert these constraints and the `Editor` default.

---

<details>
<summary><strong>Agent fix prompt</strong></summary>

```text
Target: `openapi/openapi.yaml:8343`
Affected lines: `8343-8367`, `8702-8715`

You are working in this repository. Align the access-key OpenAPI schema with the Pydantic validators.

Problem:
`AccessKeyCreateRequest` and `AccessKeyWorkspaceGrant` reject blank/invalid scope, workspace, and roles values, but `openapi/openapi.yaml`, `openapi/ga/openapi.yaml`, and `openapi/ga/individual/platform.openapi.yaml` do not publish those constraints or the default `Editor` role.

Required changes:
- Add schema constraints/default metadata to the Pydantic fields or schema source.
- Regenerate all affected OpenAPI files consistently.
- Add/update tests that assert the generated schema exposes these constraints.

Constraints:
- Preserve existing server-side normalization behavior.
- Keep the OpenAPI changes limited to the access-key request/response schemas.
```

</details>

File: packages/nemo_platform_ext/src/nemo_platform_ext/cli/commands/auth.py [hunk 3]
```diff
@@ -90,6 +93,39 @@ def _parse_access_key_expires_in(value: str | None) -> tuple[bool, int | None]:
     return True, expires_in_seconds
 
 
+def _parse_access_key_scope(value: str | None) -> list[str] | None:
+    if value is None:
+        return None
+    return list(dict.fromkeys(service.strip() for service in value.split(",") if service.strip()))
```
Comment (ironcommit) [inline packages/nemo_platform_ext/src/nemo_platform_ext/cli/commands/auth.py]: **Finding 3: Minor Blank CLI Values Surface As Unexpected Errors**

**Location:** `packages/nemo_platform_ext/src/nemo_platform_ext/cli/commands/auth.py:96`
**Affected lines:** `96-126`
**Severity:** Minor

---

**Evidence:** `_parse_access_key_scope` filters blank comma entries and can return `[]` for `--scope " , "`; `AccessKeyCreateRequest(scope=[])` then raises a Pydantic `ValidationError`. `_parse_access_key_workspace_grants` strips the workspace only when constructing `AccessKeyWorkspaceGrant`, so values such as `--workspace ":Viewer"` or whitespace-only workspace names also raise Pydantic validation errors after Typer option parsing. These are user input errors and should be `typer.BadParameter` with the relevant option hint.

**Impact:** The CLI reports these as generic unexpected errors with exit code 1 instead of clean option errors with exit code 2. The command already validates `team-a:` with `BadParameter`, so blank scope/workspace handling is inconsistent and harder for users to correct.

**Reviewed diff:**

```diff
@@
+def _parse_access_key_scope(value: str | None) -> list[str] | None:
+    if value is None:
+        return None
+    return list(dict.fromkeys(service.strip() for service in value.split(",") if service.strip()))
+
+
+def _parse_access_key_workspace_grants(
+    values: list[str] | None,
+) -> list[AccessKeyWorkspaceGrant] | None:
+    if values is None:
+        return None
+
+    grants: list[AccessKeyWorkspaceGrant] = []
+    for value in values:
+        workspace, separator, roles_value = value.partition(":")
+        roles = [role.strip() for role in roles_value.split(",") if role.strip()] if separator else []
@@
+        grants.append(
+            AccessKeyWorkspaceGrant(
+                workspace=workspace.strip(),
+                roles=roles if separator else ["Editor"],
+            )
+        )
```

**Suggested fix:** Validate these two option syntaxes before constructing Pydantic models. Raise `typer.BadParameter(..., param_hint="--scope")` when the parsed services list is empty. Strip `workspace` immediately and raise `typer.BadParameter(..., param_hint="--workspace")` when it is empty. Add CLI tests for blank `--scope`, whitespace-only `--scope`, empty workspace before `:`, and whitespace-only workspace.

---

<details>
<summary><strong>Agent fix prompt</strong></summary>

```text
Target: `packages/nemo_platform_ext/src/nemo_platform_ext/cli/commands/auth.py:96`
Affected lines: `96-126`

You are working in this repository. Fix CLI validation in `packages/nemo_platform_ext/src/nemo_platform_ext/cli/commands/auth.py` around `_parse_access_key_scope` and `_parse_access_key_workspace_grants`.

Problem:
Blank `--scope` and blank `--workspace` values fall through to Pydantic model validation and are reported as unexpected errors instead of Typer option errors.

Required changes:
- Raise `typer.BadParameter` for empty parsed scope lists, with `param_hint="--scope"`.
- Strip workspace names before model construction and raise `typer.BadParameter` when empty, with `param_hint="--workspace"`.
- Add focused CLI tests proving these cases exit as option errors and do not call the access-key API.

Constraints:
- Preserve role parsing and the default `Editor` behavior for valid workspace grants.
- Keep the fix scoped to the auth access-key create command.
```

</details>

File: docs/auth/authentication/using-authentication.mdx [hunk 4]
```diff
@@ -134,6 +134,21 @@ Pass `--expires-in <seconds>` to request a specific finite lifetime. Pass
 `--expires-in none` only for deployments where the administrator has explicitly
 allowed unlimited keys.
 
+Restrict a key to specific services with `--scope`, and grant its principal
+workspace membership with `--workspace` (formatted `<workspace>` or
+`<workspace>:<role1>,<role2>`, repeatable) instead of a separate
+`nemo workspaces members create` call:
+
+```bash
+nemo auth access-keys create --name "intake-worker" --service-account service-id \
```
Comment (ironcommit) [inline docs/auth/authentication/using-authentication.mdx:145]: **Finding 4: Minor Service-Account Example Omits PlatformAdmin Requirement**

**Location:** `docs/auth/authentication/using-authentication.mdx:143`
**Affected lines:** `137-145`
**Severity:** Minor

---

**Evidence:** The new docs example uses `--service-account`, but the surrounding text does not tell readers that service-bound access-key creation requires the PlatformAdmin role. The API enforces this requirement and returns 403 for non-admin users.

**Impact:** Users following the docs can copy the example, receive a permission error, and have to infer that the service-account flag is admin-only. This is especially confusing because the example is in the general non-SDK client workflow, not an admin-only section.

**Reviewed diff:**

```diff
@@
+Restrict a key to specific services with `--scope`, and grant its principal
+workspace membership with `--workspace` (formatted `<workspace>` or
+`<workspace>:<role1>,<role2>`, repeatable) instead of a separate
+`nemo workspaces members create` call:
+
+```bash
+nemo auth access-keys create --name "intake-worker" --service-account service-id \
+  --scope intake,entities --workspace ml-team:Editor
+```
```

**Suggested fix:** Add a short note directly before or after the example saying that `--service-account` creates a service-bound key and requires PlatformAdmin. If the section is meant for non-admin users too, include a user-bound variant of the same example without `--service-account`.

---

<details>
<summary><strong>Agent fix prompt</strong></summary>

```text
Target: `docs/auth/authentication/using-authentication.mdx:143`
Affected lines: `137-145`

You are working in this repository. Update `docs/auth/authentication/using-authentication.mdx` around the scoped access-key example.

Problem:
The example uses `--service-account`, but the docs do not state that service-bound key creation requires PlatformAdmin.

Required changes:
- Add a concise note near the example that `--service-account` requires PlatformAdmin.
- Consider adding a user-bound example without `--service-account` if this section should serve non-admin users too.

Constraints:
- Follow the Fern docs style in `docs/AGENTS.md`.
- Do not hand-edit generated CLI reference docs.
```

</details>

**Review by ironcommit — COMMENTED** [review]
Codex review comments for PR #1988.

File: docs/auth/authentication/using-authentication.mdx [hunk 5]
```diff
@@ -134,6 +134,21 @@ Pass `--expires-in <seconds>` to request a specific finite lifetime. Pass
...
+  --scope intake,entities --workspace ml-team:Editor
+```
+
+Rotate a personal key (one without `--service-account`) by passing `--rotate
```
Comment (mckornfield) [inline docs/auth/authentication/using-authentication.mdx]: one without `--service-acount` is weird phrasing, maybe `one created without...` ?

File: packages/nemo_platform_ext/src/nemo_platform_ext/cli/commands/auth.py [hunk 6]
```diff
@@ -90,6 +93,39 @@ def _parse_access_key_expires_in(value: str | None) -> tuple[bool, int | None]:
...
+
+def _parse_access_key_workspace_grants(
+    values: list[str] | None,
+) -> list[AccessKeyWorkspaceGrant] | None:
+    if values is None:
+        return None
+
+    grants: list[AccessKeyWorkspaceGrant] = []
+    for value in values:
+        workspace, separator, roles_value = value.partition(":")
```
Comment (mckornfield) [inline packages/nemo_platform_ext/src/nemo_platform_ext/cli/commands/auth.py:114]: I guess in the weird case where `:` doesn't exist, does this get messed up? I think the tuple will just return the whole string

File: services/core/auth/src/nmp/core/auth/app/access_keys.py [hunk 7]
```diff
@@ -219,6 +223,29 @@ async def handle_conflict() -> bool | _Retry:
 
         return await self._retry_on_conflict(do_attempt, handle_conflict)
 
+    async def get_for_rotation(
+        self, jti: str, principal: str, *, admin_override: AdminOverride | None = None
+    ) -> AccessKeyEntity:
+        """Validate a `rotates` target before its replacement is minted.
```
Comment (mckornfield) [inline services/core/auth/src/nmp/core/auth/app/access_keys.py]: "rotates target"?

File: services/core/auth/src/nmp/core/auth/app/access_keys.py [hunk 8]
```diff
@@ -660,8 +703,83 @@ async def create_async(
...
+                        "rotated_from_jti": request.rotates,
+                    },
+                    exc_info=True,
+                )
+                # Unlike the workspace-grant case, the new key is fully usable here; the only
+                # thing that failed is retiring the old one. The old key is still active and
+                # untouched, so compensate by revoking the new key and surfacing the failure so
+                # the caller can retry rotation, leaving the untouched predecessor as the
+                # caller's one valid credential in the meantime.
+                await self._revoke_after_failed_creation(key.jti, "rotation_revoke_failed")
```
Comment (mckornfield) [inline services/core/auth/src/nmp/core/auth/app/access_keys.py:760]: so we tried to rotate, couldn't  revoke, so we revoke the new one? I guess...

File: packages/nemo_platform_ext/src/nemo_platform_ext/cli/commands/auth.py (hunk 6 above)
Comment (anastasia-nesterenko) [reply packages/nemo_platform_ext/src/nemo_platform_ext/cli/commands/auth.py:114]: @mckornfield, do you mean like this case?
```
print("team-a".partition(":"))
('team-a', '', '')
```
or something different?

File: services/core/auth/src/nmp/core/auth/app/access_keys.py (hunk 7 above)
Comment (anastasia-nesterenko) [reply services/core/auth/src/nmp/core/auth/app/access_keys.py]: updated

File: services/core/auth/src/nmp/core/auth/app/access_keys.py (hunk 8 above)
Comment (anastasia-nesterenko) [reply services/core/auth/src/nmp/core/auth/app/access_keys.py:760]: added removing the workspaces if it failed so that it gets to the initial state before failure

File: packages/nemo_platform_ext/src/nemo_platform_ext/cli/commands/auth.py (hunk 6 above)
Comment (mckornfield) [reply packages/nemo_platform_ext/src/nemo_platform_ext/cli/commands/auth.py:114]: yeah exactly hehe

**Review by mckornfield — APPROVED** [review]
just my one comment still open on the `:` case you printed, not sure if the code needs to change or if it's unreachable

File: script/generate_config_docs.py [hunk 9]
```diff
@@ -301,6 +302,24 @@ def _get_backend_config_models(model_class: type, field_name: str) -> dict[str,
     return None
 
 
+def _model_construct_no_env(model_class: type) -> Any:
+    """``model_construct()``, but also rebuilds any BaseSettings-typed fields via
```
Comment (mckornfield) [inline script/generate_config_docs.py]: so weird but 🤷

**Review by ironcommit — APPROVED** [review]
LGTM
